pkg/controller/kubedirectorcluster/cluster.go

```diff
@@ -57,6 +57,65 @@ func (r *ReconcileKubeDirectorCluster) syncCluster(
 		cr.Status = &kdv1.KubeDirectorClusterStatus{}
 		cr.Status.Roles = make([]kdv1.RoleStatus, 0)
 	}
+
+	// Check, if some statefulset is in the middle of upgrade/rollback process.
+	// If so, the current cluster state still can be in `configured` state, change it to `updating`
+	upgradeIsActive := false
+	for _, rs := range cr.Status.Roles {
+		if rs.RoleUpgradeStatus == kdv1.RoleRollingBack || rs.RoleUpgradeStatus == kdv1.RoleUpgrading {
+			upgradeIsActive = true
+			cr.Status.State = string(clusterUpdating)
+			break
+		}
+	}
+
+	// If all statefulsets completed their upgrade/rollback processes
+	// erase the cluster upgradeInfo object
+	if cr.Status.UpgradeInfo != nil && !upgradeIsActive {
+		// If there wasn't rollback
+		if cr.Spec.AppID != cr.Status.UpgradeInfo.PrevApp {
+			// Remove the cluster bind with the previous app
+			shared.RemoveClusterAppReference(
+				cr.Namespace,
+				cr.Name,
+				*cr.Spec.AppCatalog,
+				cr.Status.UpgradeInfo.PrevApp,
+			)
+			// And bind the cluster with the new app
+			shared.EnsureClusterAppReference(
+				cr.Namespace,
+				cr.Name,
+				*cr.Spec.AppCatalog,
+				cr.Spec.AppID,
+			)
+		}
+		cr.Status.UpgradeInfo = nil
```

Comment:
All roleUpgradeStatus and podUpgradeStatus should also be cleared when the upgrade/rollback is complete, yeah?

Reply:
Don't we need to have the permanently stored info in the cluster status, that some pods / statefulsets were upgraded or rolledBack?